backend/database/bespoke.ts

```diff
@@ -180,4 +182,65 @@ export class BespokeQueries {
 
     return value[0][Object.keys(value[0])[0]];
   }
+
+  static async getReturnBalanceItemsQty(
+    db: DatabaseCore,
+    schemaName: string,
+    docName: string
+  ): Promise<DocValueMap[] | undefined> {
+    const docItems = (await db.knex!(`${schemaName}Item`)
+      .select('item')
+      .where('parent', docName)
+      .groupBy('item')
+      .sum({ quantity: 'quantity' })) as DocValueMap[];
+
+    const returnDocNames = (
+      await db.knex!(schemaName)
+        .select('name')
+        .where('returnAgainst', docName)
+        .andWhere('submitted', true)
+        .andWhere('cancelled', false)
+    ).map((i: { name: string }) => i.name);
+
+    if (!returnDocNames.length) {
+      return;
+    }
+
+    const returnedItems = (await db.knex!(`${schemaName}Item`)
+      .select('item')
+      .sum({ quantity: 'quantity' })
+      .whereIn('parent', returnDocNames)
+      .groupBy('item')) as DocValueMap[];
+
+    if (!returnedItems.length) {
+      return;
+    }
+
+    const returnBalanceItemQty = [];
+
+    for (const item of returnedItems) {
+      const docItem = docItems.filter(
```

Comment:
This is bad code, you've created an `O(n^2)` piece of code where `O(n)` would've sufficed.

`filter` loops over all the `docItems` to find matching items.

Another bad thing is you've considered only the first item, what if there are several rows for a single item each pertaining to a different batch?